driver-core/src/main/java/com/datastax/driver/core/Requests.java

```diff
@@ -240,9 +241,13 @@ public String toString() {
 
         static void serialize(EnumSet<QueryFlag> flags, ByteBuf dest, ProtocolVersion version) {
             int i = 0;
-            for (QueryFlag flag : flags)
-                i |= 1 << flag.ordinal();
-            if (version.compareTo(ProtocolVersion.V5) >= 0) {
+            boolean isV5 = version.compareTo(ProtocolVersion.V5) >= 0;
+            for (QueryFlag flag : flags) {
+                // don't include flag if would make value width greater than int and protocol < 5
+                if (flag.ordinal() < QueryFlag.KEYSPACE.ordinal() || isV5)
```

Comment:
One oddity, if you set the keyspace flag and using protocol v4, C* 4.0+ will fail since it will still attempt to parse the keyspace.   Because of this figured i'd add a check here to not add the flag if protocol is less than v5.

Reply:
Should we open a C* ticket about it? Also maybe we can add a warning here in case the user sets a keyspace and expects it to work, but for some reason (let's say protocol negotiation) the driver ends up using protocol V4.

Reply:
I didn't realize this at first, but apparently this is the way cassandra handles all query flags: 

https://github.com/apache/cassandra/blob/6bea7de50418a7a7b9e2b8dcf2446701d9cc5b69/src/java/org/apache/cassandra/cql3/QueryOptions.java#L442

If a flag is present, it will attempt to parse it, regardless of protocol version.  I think it leaves it up to the driver to behave appropriately and only set flags and values that are supported by the protocol version in use.  Not sure if that should be considered a bug, and if it was I doubt it would be fixed.

The main issue is we can't really determine whether or not the keyspace was set for routing, or to be used on the query (since they use the same method `setKeyspace`).  With other query features, we can preemptively throw an error if a user tries using a feature that isn't supported by their protocol version.   Because we can't tell why `setKeyspace` was used I think adding a warning might be more annoying for users than not.  Having queries fail is just as good of a solution I think.

Reply:
> Having queries fail is just as good of a solution I think.

Wdym having queries fail? It sounded like here the keyspace would just be ignored in the query flags if protocol_version < 5. Would the driver fail if protocol < 5 and the keyspace is set or you did you mean the query would fail because supposedly the keyspace wouldn't exist and cassandra would error out?

Reply:
For < protocol V5, the keyspace flag isn't set at all.  So the query would be executed and fail since the wrong keyspace is used (unless there is an equivalent enough schema).